parser/lexer.go

```diff
@@ -0,0 +1,499 @@
+					tok, lit = bitLit, s.r.data(&pos)
+				} else {
+					tok = unicode.ReplacementChar
+				}
+				return
+			}
+			s.r.p = pos
+		}
+		return s.scanIdent()
+	} else if isDigit(ch0) || ch0 == '.' {
+		return s.scanNumber()
+	} else if ch0 == '\'' || ch0 == '"' {
+		return s.scanString()
+	} else if ch0 == '`' {
+		return s.scanQuotedIdent()
+	}
+
+	switch ch0 {
+	case '@':
+		return s.startWithAt()
+	case '/':
+		return s.startWithSlash()
+	case '-':
+		return s.startWithDash()
+	case '#':
+		s.r.incAsLongAs(func(ch byte) bool {
+			return ch != '\n'
+		})
+		return s.scan()
+	}
+
+	// search a trie to scan a token
+	ch := ch0
+	node := &ruleTable
+	for {
+		if node.childs[ch] == nil || s.r.eof() {
+			break
+		}
+		node = node.childs[ch]
+		s.r.inc()
+		ch = s.r.peek()
+	}
+
+	tok, lit = node.token, s.r.data(&pos)
+	return
+}
+
+func (s *Scanner) startWithDash() (tok int, pos Pos, lit string) {
+	pos = s.r.pos()
+	if !strings.HasPrefix(s.r.s[pos.Offset:], "-- ") {
+		tok = int('-')
+		s.r.inc()
+		return
+	}
+
+	s.r.incN(3)
+	s.r.incAsLongAs(func(ch byte) bool {
+		return ch != '\n'
+	})
+	return s.scan()
+}
+
+func (s *Scanner) startWithSlash() (tok int, pos Pos, lit string) {
+	pos = s.r.pos()
+	s.r.inc()
+	ch0 := s.r.peek()
+	if ch0 == '*' {
+		s.r.inc()
+		for {
+			ch0 = s.r.readByte()
+			if ch0 == 0 && s.r.eof() {
+				tok = unicode.ReplacementChar
+				return
+			}
+			if ch0 == '*' && s.r.readByte() == '/' {
+				break
+			}
+		}
+		return s.scan()
+	}
+	tok = int('/')
+	return
+}
+
+func (s *Scanner) startWithAt() (tok int, pos Pos, lit string) {
+	pos = s.r.pos()
+	s.r.inc()
+	ch1 := s.r.peek()
+	if isIdentFirstChar(ch1) {
+		s.scanIdent()
+		tok, lit = userVar, s.r.data(&pos)
+	} else if ch1 == '@' {
+		s.r.inc()
+		stream := s.r.s[pos.Offset+2:]
+		for _, v := range []string{"global.", "session.", "local."} {
+			if strings.HasPrefix(stream, v) {
+				s.r.incN(len(v))
+				break
+			}
+		}
+		s.scanIdent()
+		tok, lit = sysVar, s.r.data(&pos)
+	} else {
+		tok = at
+	}
+	return
+}
+
+func (s *Scanner) scanIdent() (tok int, pos Pos, lit string) {
+	pos = s.r.pos()
+	s.r.incAsLongAs(isIdentChar)
+	return identifier, pos, s.r.data(&pos)
+}
+
+func (s *Scanner) scanQuotedIdent() (tok int, pos Pos, lit string) {
+	pos = s.r.pos()
+	s.r.inc()
+	s.buf.Reset()
+	for {
+		ch := s.r.readByte()
+		if ch == 0 && s.r.eof() {
+			tok = unicode.ReplacementChar
+			return
+		}
+		if ch == '`' {
+			if s.r.peek() != '`' {
+				tok, lit = identifier, s.buf.String()
+				return
+			}
+			s.r.inc()
+		}
+		s.buf.WriteByte(ch)
+	}
+}
+
+func (s *Scanner) scanString() (tok int, pos Pos, lit string) {
+	s.buf.Reset()
+	tok, pos = stringLit, s.r.pos()
+	ending := s.r.readByte()
+	for {
+		if s.r.eof() {
+			tok, lit = 0, s.buf.String()
+			return
+		}
+		ch0 := s.r.peek()
+		if ch0 == ending {
+			s.r.inc()
+			if s.r.peek() != ending {
+				break
+			}
+		}
+		// TODO this would break reader's line and col information
+		if ch0 == '\n' {
```

Comment:
I'll do it next PR.
scanString should be improved: more clear, confirm to standard SQL stringLit, and avoid allocation if possible.